crates/extensions/c8y_auth_proxy/src/server.rs

```diff
@@ -135,23 +165,131 @@ impl FromRef<AppState> for SharedTokenManager {
+    match message {
+        In::Text(t) => Out::Text(t),
+        In::Binary(t) => Out::Binary(t),
+        In::Ping(t) => Out::Ping(t),
+        In::Pong(t) => Out::Pong(t),
+        In::Close(Some(InCf { code, reason })) => Out::Close(Some(OutCf {
+            code: code.into(),
+            reason,
+        })),
+        In::Close(None) => Out::Close(None),
+        In::Frame(_) => unreachable!("This function is only called when reading a message"),
+    }
+}
+
+async fn connect_to_websocket(
+    token: &str,
+    headers: &HeaderMap<HeaderValue>,
+    uri: &str,
+) -> Result<WebSocketStream<MaybeTlsStream<TcpStream>>, tokio_tungstenite::tungstenite::Error> {
+    let mut req = Request::builder();
+    for (name, value) in headers {
+        req = req.header(name.as_str(), value);
+    }
+    req = req.header("Authorization", format!("Bearer {token}"));
+    let req = req
+        .uri(uri)
+        .body(())
+        .expect("Builder should always work as the headers are copied from a previous request, so must be valid");
+    tokio_tungstenite::connect_async(req)
+        .await
+        .map(|(c8y, _)| c8y)
+}
 
-impl fmt::Display for TargetHost {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        self.0.fmt(f)
+async fn proxy_ws(
+    ws: WebSocket,
+    host: TargetHost,
+    retrieve_token: State<SharedTokenManager>,
+    headers: HeaderMap<HeaderValue>,
+    path: String,
+) {
+    let uri = format!("{}/{path}", host.ws);
+    let mut token = retrieve_token.not_matching(None).await;
+    let c8y = match connect_to_websocket(&token, &headers, &uri).await {
+        Ok(c8y) => Ok(c8y),
+        Err(_) => {
+            token = retrieve_token.not_matching(Some(&token)).await;
+            match connect_to_websocket(&token, &headers, &uri).await {
+                Ok(c8y) => Ok(c8y),
+                Err(e) => Err(anyhow::Error::from(e)),
+            }
+        }
+    }
+    .expect("Error connecting to proxied websocket");
+    let (mut to_c8y, mut from_c8y) = c8y.split();
+    let (mut to_client, mut from_client) = ws.split();
+    let mut client_to_c8y = tokio::spawn(async move {
+        while let Some(Ok(msg)) = from_client.next().await {
+            to_c8y.send(axum_to_tungstenite(msg)).await.unwrap();
+        }
+    });
+    let mut c8y_to_client = tokio::spawn(async move {
+        while let Some(Ok(msg)) = from_c8y.next().await {
+            to_client.send(tungstenite_to_axum(msg)).await.unwrap();
+        }
+    });
+
+    tokio::select! {
+        res = (&mut client_to_c8y) => {
+            if let Err(e) = res {
+                error!("Websocket error proxying messages from the client to Cumulocity: {e}");
+            }
+            c8y_to_client.abort();
+        }
+        res = (&mut c8y_to_client) => {
+            if let Err(e) = res {
+                error!("Websocket error proxying messages from Cumulocity to the client: {e}");
+            }
+            c8y_to_client.abort();
+        }
```

Comment:
Query: Since this whole function doesn't return anything even while detecting one of the above failures, how would the wrapping server react? Continue functioning without the proxying capability? Wouldn't it be better to propagate these errors so that the server can choose to shutdown/panic, with the hope of restarting without that issue?

This might be out of scope for this PR specifically, as it's related to our general strategy on how we handle partial failures of a single component, in a multi-component process.

Reply:
I'm not sure *exactly* what axum does internally, but my belief is that this is a separate task spawned when we call [`ws.on_upgrade`](https://docs.rs/axum/0.6.20/axum/extract/ws/struct.WebSocketUpgrade.html#method.on_upgrade). In any case, it only affects the current connection, nothing else, and I believe this is the most desirable behaviour. If we were talking over HTTP, this is a point at which I would give up and respond with `500 Internal Server Error` to the current request, but I wouldn't want the server to stop running, particularly as the issue at this point is quite likely one with Cumulocity.